pkg/auth/dependency/interaction/provider_action.go

```diff
@@ -0,0 +1,104 @@
+package interaction
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/identity"
+	"github.com/skygeario/skygear-server/pkg/core/skyerr"
+)
+
+func (p *Provider) PerformAction(i *Interaction, step Step, action Action) error {
+	state, err := p.GetInteractionState(i)
+	if err != nil {
+		return err
+	}
+
+	var stepState *StepState
+	for _, s := range state.Steps {
+		if s.Step == step {
+			stepState = &s
+			break
+		}
+	}
+	if stepState == nil {
+		return ErrInvalidStep
+	}
+
+	switch intent := i.Intent.(type) {
+	case *IntentLogin:
+		return p.performActionLogin(i, intent, stepState, state, action)
+	}
+	panic(fmt.Sprintf("interaction: unknown intent type %T", i.Intent))
+}
+
+func (p *Provider) performActionLogin(i *Interaction, intent *IntentLogin, step *StepState, s *State, action Action) error {
+	switch step.Step {
+	case StepAuthenticatePrimary, StepAuthenticateSecondary:
+		switch action := action.(type) {
+		case *ActionAuthenticate:
+			authen, err := p.doAuthenticate(i, step, &i.State, intent.Identity, action.Authenticator, action.Secret)
+			if errors.Is(err, ErrInvalidCredentials) {
+				i.Error = skyerr.AsAPIError(err)
+				return nil
+			} else if err != nil {
+				return err
+			}
+
+			if step.Step == StepAuthenticatePrimary {
+				i.PrimaryAuthenticator = authen
+				i.SecondaryAuthenticator = nil
+			} else {
+				i.SecondaryAuthenticator = authen
+			}
+			i.State = nil
+			i.Error = nil
+			return nil
+
+		case *ActionTriggerOOBAuthenticator:
+			// TODO(interaction): handle OOB trigger
+		default:
+			panic(fmt.Sprintf("interaction: unhandled authenticate action %T", action))
+		}
+
+	case StepSetupSecondaryAuthenticator:
+		// TODO(interaction): setup secondary authenticator
+
+	case StepCommit:
+		// TODO(interaction): allow setup bearer token
+
+	}
+	panic("interaction_login: unhandled step " + step.Step)
+}
+
+func (p *Provider) doAuthenticate(i *Interaction, step *StepState, astate *map[string]string, is IdentitySpec, as AuthenticatorSpec, secret string) (*AuthenticatorInfo, error) {
+	userID, iden, err := p.Identity.GetByClaims(is.Type, is.Claims)
+	if errors.Is(err, identity.ErrIdentityNotFound) {
+		return nil, ErrInvalidCredentials
+	} else if err != nil {
+		return nil, err
+	}
+
+	authen, err := p.Authenticator.Authenticate(userID, as, astate, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	if step.Step == StepAuthenticateSecondary {
```

Comment:
Why do we need to check if `step.Step == StepAuthenticateSecondary` only?

Reply:
`doAuthenticate` is used for `StepAuthenticatePrimary` or `StepAuthenticateSecondary`. For `StepAuthenticatePrimary`, we would not provide user with concrete authenticators, but available authenticators as configured, so there is no need to check for `StepAuthenticatePrimary`.

I'll add a comment explaining this later.